clang/lib/Driver/ToolChains/Clang.cpp

```diff
@@ -4738,9 +4761,10 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
     }
 
     // Turn on Dead Parameter Elimination Optimization with early optimizations
-    if (!(RawTriple.isAMDGCN()) &&
+    if (!(RawTriple.isAMDGCN()) && !Args.hasArg(options::OPT_fsycl_link_EQ) &&
         Args.hasFlag(options::OPT_fsycl_dead_args_optimization,
-                     options::OPT_fno_sycl_dead_args_optimization, false))
+                     options::OPT_fno_sycl_dead_args_optimization,
+                     isSYCLOptimizationO2orHigher(Args)))
```

Comment:
Another thought on the `-O` level: instead of driver-level checks, could it be cleaner to avoid scheduling the pass within `clang/lib/CodeGen/BackendUtil.cpp` upon insufficient `OptimizationLevel::` value? Especially since the driver doesn't aim to diagnose the silent cancellation of DAE whenever it's less than `O2`.

Reply:
So would this be more of an exercise of the driver passing the enabling, and the BE would cancel the enabling if optimization level is less than -O2?  What about the case in which the user wants to force it to be turned on regardless of -O0 or not?  Is that valid?

Reply:
> So would this be more of an exercise of the driver passing the enabling, and the BE would cancel the enabling if optimization level is less than -O2?

Yes, that's the essence of my suggestion.

> What about the case in which the user wants to force it to be turned on regardless of -O0 or not? Is that valid?

On whether such a scenario should be allowed, I don't know if there's a use case for that. In my understanding, the compiler users usually turn off the optimizations when they aim for better debuggability and compile-time performance. Can't say if our DAE implementation impacts the debuggability much - it probably doesn't - yet in both instances, users are generally prepared for the worsening of the runtime performance.

With that, if there's a particular use-case that requires such a "bypass" of the optimization level, it would truly be easier to implement this with the driver-level checks of O levels. Otherwise, we'd have to throw in some additional `CodeGenOpt` or somehow change the current one from `bool`-like to `enum`-like. But in any case, the current changes don't seem to allow such use-cases yet, and some CLI extension would probably be required (explicit passing of `-fsycl-dead-args-optimization` == enabling it always? additional flag? additional value `always` to `-fsycl-dead-args-optimization=`?).

To sum up, if there's a known use-case, let's enable it, and if there's none, my initial suggestion stands!

Reply:
We can look at this like any other optimization.  Some have abilities to disable explicitly, and some don't.  I think the ideal way to approach this is to have this optimization be turned on (without an implicit switch, based on optimization level in `CodeGenOpt`.  We would add a disabling option, which would be hooked to `-fno-sycl-dead-args-optimization` and have a 'forced' enabling if `-fsycl-dead-args-optimization` is passed on the command line.

Perhaps we can do that transform in the future, and keep this one as is, adding a TODO

Reply:
> What about the case in which the user wants to force it to be turned on regardless of -O0 or not? Is that valid?

I don't think DAE will work with `-O0`. To make it work we need at least remove dead stores i.e. to enable DAE other optimizations must be executed first.